lib/rtti.h

```diff
@@ -0,0 +1,313 @@
+struct TypeInfo {
+    using T = std::remove_const_t<std::remove_reference_t<This>>;
+
+    static_assert((... && std::is_base_of_v<Parents, This>),
+                  "One or more parents are not a base of this type.");
+
+    static_assert((... && std::is_base_of_v<Base, Parents>),
+                  "One or more parent hierarchies is not based on top of RTTI::Base.");
+
+    static_assert(std::is_base_of_v<Base, This>,
+                  "Invalid typeinfo requested, class is not based on top of RTTI::Base.");
+
+    static_assert(
+        std::is_same_v<T *, decltype(T::rttiEnabledMarker(nullptr))>,
+        "Invalid typeinfo requested, class does not declare typeinfo via DECLARE_TYPEINFO.");
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
+    }
+};
+
+struct Base {
+    virtual ~Base() = default;
+
+    /// Queries information about dynamic type of an object.
+    /// @return typeid of current object.
+    [[nodiscard]] virtual TypeId typeId() const noexcept = 0;
+    /// Checks if given object is of specified dynamic type.
+    /// @param typeid to check with
+    [[nodiscard]] virtual bool isA(TypeId typeId) const noexcept = 0;
+
+    /// Checks if given object is of specified dynamic type.
+    /// Same as `isA`, but typeid to check with is derived from template argument.
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept {
+        return isA(TypeInfo<T>::id());
+    }
+
+    /// Casts current object to a given type `T`.
+    /// Overall the semantics is similar to that of `dynamic_cast`, however,
+    /// relies on custom RTTI metadata and intrusive annotation.
+    /// @return nullptr if object is not compatible with `T` (e.g. if `T` is not
+    /// among its base classes), casted pointer to object of T* type otherwise.
+    template <typename T>
+    [[nodiscard]] T *to() noexcept {
+        return reinterpret_cast<T *>(const_cast<void *>(toImpl(TypeInfo<T>::id())));
+    }
+
+    /// Same as `to`, but returns const pointer to T.
+    template <typename T>
+    [[nodiscard]] const T *to() const noexcept {
+        return reinterpret_cast<const T *>(toImpl(TypeInfo<T>::id()));
+    }
+
+ protected:
+    [[nodiscard]] virtual const void *toImpl(TypeId typeId) const noexcept = 0;
+};
+
```

Comment:
Might be useful to add

    template <typename T>
    static inline const T *to(const Base *ptr) noexcept {
        return ptr ? ptr->to<T>() : nullptr;
    }

That way, most `dynamic_cast` calls (eg, those in all the `Visitor::visit` functions) could be replaced with `RTTT::to`

Reply:
We already replaced these in `Visitors`, maybe I missed something?